freqtrade/data/dataprovider.py

```diff
@@ -63,9 +79,100 @@ def _set_cached_df(
+
+        :param pair: pair to get the data for
+        :param timeframe: Timeframe to get data for
+        :param candle_type: Any of the enum CandleType (must match trading mode!)
+        """
+        pair_key = (pair, timeframe, candle_type)
+
+        # For multiple leaders, if the data already exists, we'd merge
+        self.__external_pairs_df[pair_key] = (dataframe, datetime.now(timezone.utc))
+        self._set_data_event(pair_key)
+
+    def get_external_df(
+        self,
+        pair: str,
+        timeframe: str,
+        candle_type: CandleType
+    ) -> DataFrame:
+        """
+        Get the pair data from the external sources. Will wait if the policy is
+        set to, and data is not available.
+
+        :param pair: pair to get the data for
+        :param timeframe: Timeframe to get data for
+        :param candle_type: Any of the enum CandleType (must match trading mode!)
+        """
+        pair_key = (pair, timeframe, candle_type)
+
+        if pair_key not in self.__external_pairs_df:
+            self._wait_on_data(pair_key)
+
+            if pair_key not in self.__external_pairs_df:
+                return (DataFrame(), datetime.fromtimestamp(0, tz=timezone.utc))
+
+        return self.__external_pairs_df[pair_key]
+
+    def _set_data_event(self, key: PairWithTimeframe):
+        """
+        Depending on the WaitDataPolicy, if an event exists for this PairWithTimeframe
+        then set the event to release main thread from waiting.
+
+        :param key: PairWithTimeframe
+        """
+        pair_event = self.__external_pairs_event.get(key)
+
+        if pair_event:
+            num_concat, event = pair_event
+            self.__external_pairs_event[key] = (num_concat + 1, event)
+
+            if self._wait_data_policy == WaitDataPolicy.one:
+                logger.debug("Setting Data as policy is One")
+                event.set()
+            elif self._wait_data_policy == WaitDataPolicy.all and num_concat == self._num_sources:
+                logger.debug("Setting Data as policy is all, and is complete")
+                event.set()
+
+                del self.__external_pairs_event[key]
+
+    def _wait_on_data(self, key: PairWithTimeframe):
+        """
+        Depending on the WaitDataPolicy, we will create and wait on an event until
+        set that determines the full amount of data is available
+
+        :param key: PairWithTimeframe
+        """
+        if self._wait_data_policy is not WaitDataPolicy.none:
+            pair, timeframe, candle_type = key
+
+            pair_event = Event()
+            self.__external_pairs_event[key] = (0, pair_event)
+
+            timeout = self._wait_data_timeout \
+                if self._wait_data_policy is not WaitDataPolicy.all else 0
+
+            timeout_str = f"for {timeout} seconds" if timeout > 0 else "indefinitely"
+            logger.debug(f"Waiting for external data on {pair} for {timeout_str}")
```

Comment:
How does this Event class work?
will i only be able to "get" from it once - or forever once the data is available at least once?

We'll be needing the latter case. 
You can store the data with a timestamp of analysis and reject old data - but you must assume that `get_external_df()` can be called multiple times per candle - and will have to return the same object each time until a new dataframe arrives.
the assumption that the strategy analyzes a pair only once per candle is true for freqAI - but is not generally applicable - and therefore a wrong assumption.

As such - i can have a strategy that's analyzing once every loop - adding "live" data to the leader's data. 
That's perfectly valid - but the way i understand the "Event" class - would block my loop until the next candle arrives.

Reply:
As of right now, during `Strategy.analyze_external` it will use the data from `get_external_df` and run it through `populate_entry/exit_signal` as if it had just come from `populate_indicators`. The flow after that is the same as the normal `process` execution where it caches it in the DataProvider using `_set_cached_df` to be referenced later. I think a way of invalidating the age of the data would be to delete it as soon as it's used. However, when the dataframe is added to the `DataProvider` by the `ExternalMessageConsumer` it is marked with a timestamp to indicate that is when the dataframe was received/calculated.

Reply:
why use Event and not write directly to the cache once the data is arriving? 
You can wrap it in a simple threading.Lock - and it'll be threadsafe, too.

Using event here seems overly complex and unnecessary - it's adding 2 full functions which are unnecessary (especially since we will not be waiting for the dataframe to maybe arrive).